engine/time/src/main/java/io/deephaven/time/calendar/BusinessCalendar.java

```diff
@@ -57,76 +57,176 @@ public InvalidDateException(final String message, final Throwable cause) {
+            if (month == 1) {
+                startInstant = ms.startInstant;
+                startDate = ms.startDate;
+            }
 
-        final int yearStart =
-                firstValidDate.getDayOfYear() == 1 ? firstValidDate.getYear() : firstValidDate.getYear() + 1;
-        final int yearEnd = ((lastValidDate.isLeapYear() && lastValidDate.getDayOfYear() == 366)
-                || lastValidDate.getDayOfYear() == 365) ? lastValidDate.getYear() : lastValidDate.getYear() - 1;
+            if (month == 12) {
+                endInstant = ms.endInstant;
+                endDate = ms.endDate;
+            }
 
-        for (int year = yearStart; year <= yearEnd; year++) {
-            final LocalDate startDate = LocalDate.ofYearDay(year, 1);
-            final LocalDate endDate = LocalDate.ofYearDay(year + 1, 1);
-            final ZonedDateTime start = startDate.atTime(0, 0).atZone(timeZone());
-            final ZonedDateTime end = endDate.atTime(0, 0).atZone(timeZone());
+            businessTimeNanos += ms.businessTimeNanos;
+            businessDays += ms.businessDays;
+            nonBusinessDays += ms.nonBusinessDays;
+            businessDates.addAll(ms.businessDates);
+            nonBusinessDates.addAll(ms.nonBusinessDates);
+        }
 
-            LocalDate date = startDate;
-            long businessTimeNanos = 0;
+        return new SummaryData(
+                year,
+                startInstant,
+                startDate,
+                endInstant,
+                endDate,
+                businessTimeNanos,
+                businessDays,
+                nonBusinessDays,
+                businessDates,
+                nonBusinessDates);
+    }
 
-            while (date.isBefore(endDate)) {
-                final CalendarDay<Instant> bs = this.calendarDay(date);
-                businessTimeNanos += bs.businessNanos();
-                date = date.plusDays(1);
-            }
+    private SummaryData getYearSummary(final int year) {
 
-            final YearData yd = new YearData(start.toInstant(), end.toInstant(), businessTimeNanos);
-            cachedYearData.put(year, yd);
+        if (year < yearCacheStart || year > yearCacheEnd) {
+            throw new InvalidDateException("Business calendar does not contain a complete year for: year=" + year);
         }
+
+        return summaryCache.getYearSummary(year);
     }
 
-    private YearData getYearData(final int year) {
-        final YearData yd = cachedYearData.get(year);
 
-        if (yd == null) {
-            throw new InvalidDateException("Business calendar does not contain a complete year for: year=" + year);
+    private ImmutableConcurrentCache.Pair<CalendarDay<Instant>> computeCalendarDay(final int yearMonthDay) {
+        final int year = yearMonthDay / 10000;
+        final int monthDay = yearMonthDay % 10000;
+        final int month = monthDay / 100;
+        final int day = monthDay % 100;
```

Comment:
Ditto with stuffing all this stuff into an int. I'm generally concerned about it. But if we are going to do it, would should colocate the generator of extractor for them.

Reply:
I pushed some changes that are a compromise.  They should minimize boxing and garbage and should consolidate logic.  To clean it up further would probably require worse performance.